06.bowling_object/lib/game.rb

```diff
@@ -10,25 +10,17 @@ def initialize(shot_pins)
   end
 
   def score
-    @frames.each_with_index.sum { |frame, idx| frame.score(@frames, idx) }
+    @frames.sum { |frame| frame.score(@frames) }
   end
 
   private
 
   def build_frames(shot_pins)
-    @frames = []
-    shot_pins.split(',').each do |pin|
-      @frames = [*@frames, Frame.new] if frame_terminated?(@frames.last) && !last_frame?
-      @frames.last.shots = @frames.last.add(pin)
+    frames = []
+    shot_pins.split(',').each_with_index do |pin, idx|
+      frames = [*frames, Frame.new(frames.size)] if idx.zero? || frames[-1].two_shots_done? && !frames[-1].last_frame?
```

Comment:
```suggestion
      frames << Frame.new(frames.size) if idx.zero? || frames[-1].two_shots_done? && !frames[-1].last_frame?
```

プログラミング言語によっては配列への要素追加を許さない（＝常にイミュータブルな）言語もありますが、Rubyではそういう制限は無いので `<<` や `push` を使った方がシンプルです。

Reply:
あと、僕だったらこんなふうにして「現在のループで参照しているフレーム」を`current_frame`として扱えるようにしますね。

```ruby
  def build_frames(shot_pins)
    frames = []
    current_frame = nil
    shot_pins.split(',').each do |pin|
      if current_frame.nil? || current_frame.two_shots_done? || !current_frame.last_frame?
        current_frame = Frame.new(frames.size)
        frames << current_frame
      end
      current_frame.add(pin)
    end
    frames
  end
```

Reply:
nil で初期化すればおさまりよくなるんですね...修正しました